dot/network/notifications.go

```diff
@@ -270,6 +270,7 @@ func (s *Service) sendData(peer peer.ID, hs Handshake, info *notificationsProtoc
 		return
 	}
 
+	info.peersData.setMutex(peer)
```

Comment:
The method `sendData` is been called by `broadcastExcluding` which also sets the mutex, I would suggest to set the mutex before calling `sendData`

Reply:
I don't understand this `setMutex` function.  It has a defer within it unlock the mutex, but it will just get unlocked as soon as this call is done.

Reply:
From what I understand `setMutex` function should be placed when we connect to a peer and only there, setting a mutex we avoid multiple network calls be executed concurrently for the same protocol. So if we have peer A sending us 3 block announces and we still processing the first one then 2 block announces will be locked waiting. However I see that we're using `setMutex` in places where it should not exist, e.g before sending a message, we should not `setMutex` here but use the `info.peersData.getMutex(peer)` and with the mutex then lock the resource, @timwu20 what do you think?

Reply:
From the codebase I see that
- Currently we are using this peer mutex only while sending a handshake. so, we can't send multiple handshakes of the same type to a particular peer simultaneously.
- we don't use this mutex while handling an incoming handshake.
- we don't use this mutex while writing to stream.

I feel like we should should use this mutex on handling incoming handshakes as well, to make sure that we don't try to create multiple connection with the same peer with the same protocol.

It should be safe to not use this mutex while writing to stream (since mutex will stop us from making same type of connections)


about where to put setMutex method I like how, I think it should happen like this https://github.com/ChainSafe/gossamer/blob/972438112f4eddd8d281b358d1b1fae31fd70517/dot/network/service.go#L299-L316

This makes me realize that I should register these two parachain protocols such that  above linked code could run. That would be the ideal solution.

I bet we must be facing this setMutex related problems in grandpa service as well. 
So, if you go through how services start, network service starts first and then grandpa service starts. So this code of running connect handler and disconnect handler never runs for grandpa protocol.

https://github.com/ChainSafe/gossamer/blob/972438112f4eddd8d281b358d1b1fae31fd70517/lib/services/services.go#L43
https://github.com/ChainSafe/gossamer/blob/972438112f4eddd8d281b358d1b1fae31fd70517/lib/grandpa/network.go#L56

We should probably try to register all the network protocols at the same time through the same place. Would make these bugs go away. After we fix this, we would not have to set mutex anywhere else.

Reply:
Let me know if this makes sense to both of you. I will create a separate issue for that.

Reply:
yea I'm not really understanding the requirement to get and set mutexes based on peer id.  Like the `setMutex` function doesn't even check if a mutex is already in the map.  I feel like we should revisit the `peersData` use patterns to identify the correct solution in a new issue.

Reply:
> Like the setMutex function doesn't even check if a mutex is already in the map.

Fixable

>  I feel like we should revisit the peersData use patterns to identify the correct solution in a new issue.

Yes, Sure.

Reply:
> yea I'm not really understanding the requirement to get and set mutexes based on peer id.

I would think that this comment would explain it, but sure let's continue discussion in a separate issue and try to come up with a more appropriate solution.

https://github.com/ChainSafe/gossamer/pull/3827#discussion_r1557576846

Reply:
I have created this issue https://github.com/ChainSafe/gossamer/issues/3890. Let's provide ideas and continue discussion there.